### PR TITLE
Dispatch telemetry for a revealed password

### LINKED ISSUE #205: Implement a `PasswordReveal` action
Currently, in the `ItemDetailPresenter`, we don't dispatch an action (or listen for a `Store`-based source of truth) when hiding / showing the password. According to our [metrics document](https://github.com/mozilla-lockbox/lockbox-android/blob/master/docs/metrics.md), we should be tracking this hide / show behavior, which requires this implementation to use an `Action`-based flow rather than keeping everything local to the Presenter/View pair (which breaks Flux anyway, regardless of our metrics requirements!!)

### PR DESCRIPTION
**Pull request by linuxwolf**
Adds an action dispatched when the user would be revealing an item's password.

This patch matches the details as well as the intent.  The store is limited to just data that actual action can be taken on (password visibility), and roughly matches the implementation in iOS.

Fixes #205 

## Testing and Review Notes

Unit tests confirm the action is dispatched when revealed, and not dispatched when hidden.

**NOTE** this implementation can record a `reveal_password` telemetry event when the user shows and hides the password, ostensibly matching the behavior in iOS.

_(**Required**: steps to take to confirm this works as expected or other guidance for code, UX, and any other reviewers)_

## To Do

- [x] “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add unit tests
  - optional: consider adding instrumentation (integration/UI) tests
- consider running this branch in a debug simulator and check for memory leak notifications or any warnings
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

**Comment by linuxwolf** [comment]
@mozilla-lockbox/mobile-engineering this PR is ready for review at your convenience.

@irrationalagent I'd like your opinion on the telemetry recording; as far as I can tell, this duplicates the behavior in iOS, but it also records a `reveal_password` telemetry event each time the user taps the button -- whether it is in fact revealing or concealing the password.

File: app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    init {
+        dispatcher.register
+            .filterByType(RouteAction::class.java)
+            .subscribe { action ->
+                when (action) {
+                    is RouteAction.ItemDetail -> _passwordVisible.onNext(false)
+                    is RouteAction.ItemList -> _passwordVisible.onNext(false)
+                }
+            }
+            .addTo(compositeDisposable)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt]: AFAIUI this is "each time you go from item list to item detail or back again, make the password invisible."

Is there a simpler way we can achieve this?

File: app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt]: Nit: whitespace

File: app/src/main/java/mozilla/lockbox/action/ItemDetailAction.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/action/ItemDetailAction.kt]: Nit: whitespace.

File: app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt (hunk 1 above)
Comment (linuxwolf) [reply app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt]: nothing else was obvious.  Also note that this pattern is almost exactly what iOS does today.

File: app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt (hunk 2 above)
Comment (linuxwolf) [reply app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt]: oops!
Aaaand it's in the just-merged template, too.

File: app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt (hunk 1 above)
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/store/ItemDetailStore.kt]: See my comment re: `ItemListPresenter` dispatching an `ItemDetailAction.View(id)` action, which itself dispatches the `RouteAction`.

**Comment by linuxwolf** [comment]
> Looks good. Default is not visible, meaning if a client generates this event at least once they viewed the password at least once correct? That's all that's important from an analysis perspective.

Correct, it starts hidden, and the first tap makes it visible.

**Pull request merged**